dulwich/stash.py

```diff
@@ -62,7 +62,14 @@ def from_repo(cls, repo):
 
     def drop(self, index):
         """Drop entry with specified index."""
-        raise NotImplementedError(self.drop)
+        with open(self._reflog_path, "rb+") as f:
+            drop_reflog_entry(f, index, rewrite=True)
+        if len(self) == 0:
+            os.remove(self._reflog_path)
+            del self._repo.refs[self._ref]
+            return
+        if index == 0:
+            self._repo.refs[self._ref] = self[0].new_sha
 
     def pop(self, index):
```

Comment:
I left `pop` unimplemented since it's not entirely clear what pop should do when we aren't touching the working tree.

Reply:
today, stash doesn't update the working tree (as per its docstring) but it should.